src/scratchOrgCreate.ts

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+// third
+import { Duration } from '@salesforce/kit';
+import { SuccessResult } from 'jsforce';
+
+// Local
+import { Org } from './org';
+import { Logger } from './logger';
+import { AuthInfo } from './authInfo';
+import { Messages } from './messages';
+import { SfdxError } from './sfdxError';
+import { Connection } from './connection';
+import { SfdxProject } from './sfdxProject';
+import { Lifecycle } from './lifecycleEvents';
+import { ConfigAggregator } from './config/configAggregator';
+import {
+  authorizeScratchOrg,
+  requestScratchOrgCreation,
+  pollForScratchOrgInfo,
+  deploySettingsAndResolveUrl,
+} from './scratchOrgInfoApi';
+import SettingsGenerator from './scratchOrgSettingsGenerator';
+import { generateScratchOrgInfo, getScratchOrgInfoPayload } from './scratchOrgInfoGenerator';
+
+Messages.importMessagesDirectory(__dirname);
+const messages: Messages = Messages.loadMessages('@salesforce/core', 'scratchOrgCreate');
+
+const DEFAULT_STREAM_TIMEOUT_MINUTES = 6;
+
+export interface ScratchOrgCreateResult {
+  username?: string;
+  authInfo?: AuthInfo;
+  warnings: string[];
+}
+export interface ScratchOrgCreateOptions {
+  hubOrg: Org;
+  connectedAppConsumerKey: string;
+  durationDays: number;
+  nonamespace: boolean;
+  noancestors: boolean;
+  wait: Duration;
+  setdefaultusername: boolean;
+  setalias: string;
+  retry: number;
+  apiversion: string;
+  definitionjson: string;
+  definitionfile: string;
+  orgConfig: Record<string, unknown>;
+  configAggregator: ConfigAggregator;
+  clientSecret: string;
+}
+
+export const scratchOrgCreate = async (options: ScratchOrgCreateOptions): Promise<ScratchOrgCreateResult> => {
+  const logger = await Logger.child('scratchOrgCreate');
+  logger.debug('scratchOrgCreate');
+
+  const {
+    hubOrg,
+    connectedAppConsumerKey,
+    durationDays,
+    nonamespace,
+    noancestors,
+    wait = Duration.minutes(DEFAULT_STREAM_TIMEOUT_MINUTES),
+    setdefaultusername,
+    setalias,
+    retry = 0,
+    apiversion,
+    definitionjson,
+    definitionfile,
+    orgConfig,
+    configAggregator,
+    clientSecret,
+  } = options;
+
```

Comment:
I think we need to validate the inputs.  The command flags does a lot for us (ex: min/max on days) but this API is going to accept the inputs and won't have errors until they fail on the server.